### ISSUE TITLE
Implement UDP for mouse move messages

### ISSUE THREAD

**Issue by the-wes** [symless/synergy-core#4269]
Problem: Some users experience laggy mouse movements on slow networks

Solution: Use UDP for mouse messages instead of TCP

UDP is faster than TCP. It does this by gutting the parts of TCP which make TCP more reliable: delivery assurance, stateful connections. UDP would be ideal for mouse movements, since the next movement always overrides the previous one, making dropped packets irrelevant. However it would be less ideal for keyboard events, as a dropped packet here would mean losing keystrokes.

In short, we want keyboard events to be queued up, while mouse events not so much.

**Comment by moben** [comment]
One thing to keep in mind is that new mouse movements only override old ones when relativeMouseMoves isn't set to true. You'll also have to account for packets arriving out of order, to prevent the cursor from jumping around.

**Comment by beporter** [comment]
It seems to me that you kind of _always_ want mouse and keyboard events to occur in order. If I click in a text area, type some words, move the mouse, click in another textarea, type some more and then click (using the mouse) a Submit button all (very) quickly, it definitely matters that the text entry happens in between the mouse movements and click events. I don't want the click event for the Submit button to happen before the typing has finished, for example.

Now granted, we're talking about millisecond-to-microsecond precision that will _typically_ be faster than  human, but Synergy should be able to guarantee that **both** text and mouse input order is respected and "interleaved" properly, otherwise it can't be trusted to control a client machine with the same fidelity that a hardware keyboard and mouse can control the host machine. This would seem to **mandate** TCP in my mind.

:-1: from me on this, fwiw.

**Comment by nbolton** [comment]
@beporter I guess we could implement it as an option (disabled by default) to see what happens.

**Comment by beporter** [comment]
@nbolton That's true. Personally I'd rather see the focus remain on stability instead of experimental features where the payoff seems low (it's a reasonable assumption that a group of Synergy server/clients are always going to be on the same LAN where latency is acceptable good/stable/reliable enough for TCP.) Of course this won't_always_ be the case (otherwise this issue wouldn't have been opened) and people don't always have control over their network conditions, but situations like that are beyond reasonable control _already_. Asking Synergy to adapt to those specific cases seems unfair. 

If anything, I'd put that same research time towards [QoS](https://en.wikipedia.org/wiki/Quality_of_service#IP_and_Ethernet_efforts) and somehow signaling to the network that Synergy packets have a justifiable priority over other traffic. That and continuing to improve the  cpu footprint of synegyc and synergys, of course, I think would produce far better gains for a far larger group of users than UDP would with its obvious limitations and drawbacks for this application.

All that said-- ¯\\_(ツ)_/¯ it's your call.

**Comment by nbolton** [comment]
You'll be glad to hear, right now stability is a higher priority than UDP, by a long shot.

**Comment by sk-ws-stan** [comment]
+1 
It should be worthwhile testing sending the mouse position via UDP, but all other mouse events via TCP

**Comment by DiThi** [comment]
In my case, TCP is super laggy sometimes, losing some mouse movement messages shouold just skip a bit instead of making it laggy for a couple of seconds. Switched from wifi to bluetooth and there's no occasional congestion but more uniform packet loss so it's always laggy.

If we use UDP for mouse move only, it should also send one mouse move via TCP before every key and button event, in case the equivalent UDP packet didn't arrive and the user doesn't click where they predict. It's important for keys too, as there's places sensitive to mouse position (e.g. X-Window style focus, Blender areas).

Of course all this only would work well for absolute mouse positioning. For when relative is needed, the host should send "absolute-like" coordinates that are converted to relative by the client.

**Comment by RedFox20** [comment]
Synergy needs UDP support really badly. There are ways to implement reliable UDP that has less latency and connection issues: http://gafferongames.com/networking-for-game-programmers/reliability-and-flow-control/

The main problem with Synergy right now is constant lagging over WLAN. If a packet fails to arrive, all input is laggy because of it (or there's a problem with the network library itself). 

A really simple solution is to ID+timestamp all mouse/keyboard events, then send them over UDP 2-3 times. Yes, send the events multiple times, pre-emptively over UDP. The amount of traffic is extremely small and if your network can't deliver 3 UDP packets in time, it can't be called a reliable connection anyway. Finally, implement a 15-30ms event buffer on the client side which will order the packets and play them in a timely manner.

I'm not familiar what kind of network library is being used, but it's horrible. Synergy has been lagging the same way for over two years. What kind of systems are you testing Synergy with? It's unusable. All of this text was typed over Windows->macOS connection and it's horribly painful with 0.6s-1.5s lag for key and mouse events.

And don't blame the network setup either - I've tried Synergy over multiple gigabit networks and it lags all the same. The current setup, for example, has Windows 10 PC connected to a wifi router via LAN and a Macbook Pro connected to the same router's WLAN.

Please take a look at UDP and at least enable it as an alternative option. I've been doing low-level network programming for years and the amount of lag synergy has, is inexcusable. I hope you disabled NAGLE on the TCP socket.

**Comment by telorand** [comment]
@DiThi You said:
> If we use UDP for mouse move only, it should also send one mouse move via TCP before every key and button event...

How exactly would you accomplish that? A computer wouldn't be able to anticipate what action you are going to perform next; it can only react to what you've done. Since UDP doesn't have any built-in packet ordering, you'd have to send periodic TCP packets (based on some kind of heuristics or static frequency) to sync everything up and make sure the mouse is where it is supposed to be. Maybe this would be a viable solution, or maybe it would add some overhead such that it would be a wash; somebody smarter could probably determine that.

---

@RedFox20 I've been using Synergy for several years, most recently with SSH, and never experienced anything close to what you are describing, except when I'm launching certain programs (IntelliJ, I'm looking at you). Maybe there are other things at work in the server or client machines besides TCP issues...?

Also, that article is an interesting read. Something like that might be interesting if Symless decides to implement UDP. Thanks for sharing!

**Comment by RedFox20** [comment]
@telorand I'm really glad that you have no issues with Synergy. I wish it was the same for me. I've changed offices and computers over the past few years and the issue is always the same -- significant input lag and lag spikes that make the experience horrible.

Perhaps the only constant is that I've always used Synergy for Windows 10 <--> Macbook Pro. I'm pretty sure some part of the network stack configuration might be at fault here.

Has synergy disabled nagle (timed TCP buffering) and disabled TCP_CORK (wait until send buffer is full)?
Appropriate low-latency settings for TCP sockets should be:
TCP_NODELAY true
TCP_CORK false

**Comment by DiThi** [comment]
@telorand Just ignore UDP packets that are older than the newest packet of any type.

**Comment by telorand** [comment]
@DiThi I think that's essentially what [the article posted by @RedFox20](http://gafferongames.com/networking-for-game-programmers/reliability-and-flow-control/) talks about. It's an interesting solution.

@RedFox20 I wish I could be of more help to you. Admittedly, I usually do a Win10 <--> Win10, so that's probably why I've had fewer issues. I wonder if MacOS <--> Linux would cause the same issues, since they are both *nix systems.
